pkg/statistics/builder.go

```diff
@@ -416,13 +416,25 @@ func BuildHistAndTopN(
 		topNList = pruneTopNItem(topNList, ndv, nullCount, sampleNum, count)
 	}
 
+	topn := &TopN{TopN: topNList}
+	lenTopN := int64(len(topn.TopN))
+	if hg.NDV <= lenTopN {
+		// If we've collected everything  - don't create any buckets
+		return hg, topn, nil
+	}
+
 	// Step2: exclude topn from samples
-	if numTopN != 0 {
-		for i := int64(0); i < int64(len(samples)); i++ {
+	lenSamples := int64(len(samples))
+	if lenTopN > 0 && lenSamples > 0 {
+		for i := int64(0); i < lenSamples; i++ {
 			sampleBytes, err := getComparedBytes(samples[i].Value)
 			if err != nil {
 				return nil, nil, errors.Trace(err)
 			}
+			if i > 0 && bytes.Equal(sampleBytes, samples[i-1].Value.GetBytes()) {
```

Comment:
```suggestion
			// Safety:
			// When a sample value matches an entry in the topN list, it is removed from the samples array,
			// so it can never match the previous sample. Therefore, this equality check will never be true
			// in that case. For values not found in topN, we don't need keep checking the same value.
			if i > 0 && bytes.Equal(sampleBytes, samples[i-1].Value.GetBytes()) {
```